### ISSUE TITLE
Emit a message for reliance on variable annotations never initialized in `__init__` or `__new__`

### ISSUE THREAD

**Issue by jacobtylerwalls** [PyCQA/pylint#5806]
### Current problem

This pattern emits no messages:
```python
class PossibleNoMember:
    def __init__(self):
        self.member: bool

    def set_member(member):
        self.member = member  # this should be set in `__init__` or `__new__`

    def depend_on_member():
        print(self.member)
```

### Desired solution

Should raise either `maybe-no-member` if we're being cautious, or if we can actually check `__init__` _and_ `__new__`, then we might actually raise `no-member`.

### Additional context

See discussion at

https://github.com/PyCQA/pylint/pull/5544#issuecomment-1038768505
#2945
#3167 (fixed by #4194)

From PEP 526:
"In particular, the value-less notation a: int allows one to annotate instance variables that should be initialized in __init__ or __new__."

If a user feels strongly about initializing these variables outside `__init__` or `__new__`, I think a disable is proper. The question being whether to create a new message they could disable globally.

**Comment by Pierre-Sassoulas** [comment]
I'm +1 on this one, 

I agree with:

> If a user feels strongly about initializing these variables outside ``__init__`` or ``__new__``, I think a disable is proper.

We even have a check for this ``attribute-defined-outside-init (W0201)``

**Comment by DanielNoord** [comment]
We would need to disable this ourselves a lot, both in `pylint` and `astroid` 😅

**Comment by DanielNoord** [comment]
For `astroid` we often simply can't. There are various occasions where a `node` is created before one of its attribute nodes is created. That's why we have `postinit`.

The trouble we're running into now is that it is often much better to add the typing in the `__init__` and do the assignment in `__postinit__` than initialising as `None` and then having the put `isinstance` checks all over the place.
As we're increasing the amount of typing I would actually consider disabling this future check in `astroid`.

**Comment by DanielNoord** [comment]
That would mean that we're creating runtime costs (even it is was a singleton + couple of imports) just to satisfy `pylint`.

**Comment by Pierre-Sassoulas** [comment]
But the resulting object could be an immutable or something more efficient than what we currently have, and we're already modifying the object in postinit atm. I'm not sure that the performance cost is obvious here. But I digress, it's not a good choice for astroid for other reason (having to refactor everything) so we should disable the check (it does not mean it's a bad check in every situation though).

**Comment by jacobtylerwalls** [comment]
Let's add `__postinit__` to the allowable methods besides `__init__` and `__new__`.
